### DISCUSSION TITLE
New Lines not adding <Softbreak /> node

### DISCUSSION THREAD

**Discussion by khanakia** [yuin/goldmark#394] [Q&A] [answered]
goldmark has [https://github.com/yuin/goldmark/discussions](Discussions) in github.
You should post only issues here. Feature requests and questions should be posted at discussions.

- [X] goldmark is fully compliant with the CommonMark. Before submitting issue, you **must** read [CommonMark spec](https://spec.commonmark.org/0.29/) and confirm your output is different from [CommonMark online demo](https://spec.commonmark.org/dingus/).
    - [X] **Extensions(Autolink without `<` `>`, Table, etc) are not part of CommonMark spec.** You should confirm your output is different from other official renderers correspond with an extension.
- [X] **goldmark is not dedicated for Hugo**. If you are Hugo user and your issue was raised by your experience in Hugo, **you should consider create issue at Hugo repository at first** .

Please answer the following before submitting your issue:

1. What version of goldmark are you using? : 1.5.4
2. What version of Go are you using? : 1.20.1
3. What operating system and processor architecture are you using? : Mac
4. What did you do? : ran the below code
5. What did you expect to see? : it should add softbreak
6. What did you see instead? : not adding softbreak
7. Did you confirm your output is different from [CommonMark online demo](https://spec.commonmark.org/dingus/) or other official renderer correspond with an extension?: yes

```go
package main

import (
	"fmt"
	"github.com/yuin/goldmark"
	"github.com/yuin/goldmark/ast"
	"github.com/yuin/goldmark/text"
)

var source = `
**2. Click "Rankings Distribution"
Click "Landscape"
Click "SEO Dashboard
COMPETITIVE RESEARCH
Domain Overview
Traffic Analytics
Organic Research
Keyword Gap
Backlink Gap
KEYWORD RESEARCH
Keyword Overview..."**
`
func main() {
	doc := goldmark.DefaultParser().Parse(text.NewReader([]byte(source)))
	ast.Walk(doc, func(n ast.Node, entering bool) (ast.WalkStatus, error) {
		s := ast.WalkStatus(ast.WalkContinue)
		var err error
		fmt.Println(n.Kind(), entering, string(n.Text([]byte(source))))
		return s, err
	})
}

```

### Outputs
```
............
Text true 2. Click "Rankings Distribution"
Text false 2. Click "Rankings Distribution"
Text true Click "Landscape"
Text false Click "Landscape"
Text true Click "SEO Dashboard
Text false Click "SEO Dashboard
Text true COMPETITIVE RESEARCH
Text false COMPETITIVE RESEARCH
Text true Domain Overview
Text false Domain Overview
Text true Traffic Analytics
Text false Traffic Analytics
Text true Organic Research
Text false Organic Research
....
```

It should add the softbreak between lines print like this according to https://spec.commonmark.org/dingus/
```xml
<text>2. Click </text>
    <text>&quot;</text>
    <text>Rankings Distribution</text>
    <text>&quot;</text>
    <softbreak />
    <text>Click </text>
    <text>&quot;</text>
    <text>Landscape</text>
    <text>&quot;</text>
    <softbreak />

```

**Comment by yuin** [comment]
Spec is defined only for HTML outputs

**Comment by khanakia** [comment]
I do not understand you can see there is softbreak node the package does NOT provides.

Then how do i detect the line breaks ??

**Comment by khanakia** [comment]
@yuin can you let me know how i detect line breaks it's very critical to my use case. Thanks

**Comment by khanakia** [comment]
@yuin any update on the issue ?

**Comment by abhinav** [comment, marked as answer]
@khanakia Match on the `*ast.Text` node and check the [Text.SoftLineBreak](https://pkg.go.dev/github.com/yuin/goldmark@v1.5.4/ast#Text.SoftLineBreak) method to see if a text ends with a soft break. There's no separate soft break node in the Goldmark AST.

```go
if t, ok := n.(*ast.Text); ok {
  if t.SoftLineBreak() {
    fmt.Println("ends with softbreak")
  }
}
```
